hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -157,13 +159,17 @@ protected void onInitialize() {
     public void setCacheManager(HazelcastCacheManager cacheManager) {
         assert cacheManager instanceof HazelcastClientCacheManager;
 
-        this.cacheManager = (HazelcastClientCacheManager) cacheManager;
+        if (!cacheManagerRef.compareAndSet(null, (HazelcastClientCacheManager) cacheManager)) {
+            if (cacheManagerRef.get() != cacheManager) {
```

Comment:
Personally I would rather go for something like this:

```
@Override
    public void setCacheManager(HazelcastCacheManager cacheManager) {
        assert cacheManager instanceof HazelcastClientCacheManager;

	if(cacheManagerRef.get()==cacheManager)return;

        if (!cacheManagerRef.compareAndSet(null, (HazelcastClientCacheManager) cacheManager)) {
               throw new IllegalStateException("Cannot overwrite a Cache's CacheManager.");
        }
}
```

Since it makes more clear that you are fine with ignoring repeated settings of the same cache manager.